### PR TITLE
[TEP-0044]: Propose pipeline in pod + rollout plan

### PR DESCRIPTION
**Pull request by lbernick**
This commit proposes using Pipeline in a Pod to address TEP-0044,
and adds implementation milestones for features that should be supported in a pod.
Additional design details will be added in a subsequent commit.

### CODE REVIEW

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:191]: I've moved parallel tasks to "changes that won't require updating entrypoint" because the entrypoint already supports multiple waitfiles; it would just be a matter of updating the controller logic to determine what Tasks a given Task depends on based on a DAG.

Comment (bobcatfish) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:191]: great catch!! i was definitely overcomplicating it in my head 🤣

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (bobcatfish) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:325]: good call re. finally tasks + the parallel requirement 👍

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (bobcatfish) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:348]: maybe i missed it above but probably makes sense to include task level sidecars in the list above

i'm thinking optional but probably not that hard to add?

okay with it being required too - ive been thinking that it would be useful to be able to define a task with just a sidecar that can be easily composed with other tasks (e.g. to provide a docker in docker sidecar) - however that's a bit outside the scope of how this TEP is currently defined

**Review by bobcatfish — COMMENTED** [review]
Looks great to me!

/approve

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (afrittoli) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:335]: Does this refer to task timeouts or the whole pipeline timeout or both?
I think having at least the overall pipeline timeout is rather important, I would not want to run pipelines in a production environment if there's no control against them running indefinitely.
If timeouts are especially challenging, we could at least document available workarounds

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (afrittoli) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:336]: I think this is very important as well. In terms of timeline I agree this should come after the required features, but I think it would be sad not to have it - especially since most of the heavy lifting will be done by having support for parallel tasks and workspaces.
Of course the fallback option here is to write the tasks one by one and run the resulting pipeline in a pod :)

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (afrittoli) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:353]: Sadly yes, but having static matrix would already be a great plus I think!

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (afrittoli) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:627]: We don't have it yet in the core api - but hopefully soon enough :)

**Review by afrittoli — COMMENTED** [review]
Looks good, thank you!

I still would prefer we moved away from the "pipeline in a pod" naming of the feature - at least in the API and user docs - since the Pod is the mean and not the objective and the `Pod` is very k8s specific, but not something for this PR.

Looking forward to this, thank you @lbernick for driving it forward

/approve

**Comment by vdemeester** [comment]
Same remarks as @afrittoli but looks good to me otherwise 😉

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:348]: good call-- I've added it for now to optional features.

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:336]: makes sense! discussion in previous comment.

**Comment by lbernick** [comment]
> I still would prefer we moved away from the "pipeline in a pod" naming of the feature - at least in the API and user docs - since the Pod is the mean and not the objective and the `Pod` is very k8s specific, but not something for this PR.

Updated to note that the naming of this feature is still under discussion!

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:158]: sounds good - though would prefer that we make this decision (authoring time vs runtime) before marking this TEP implementable

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:311]: this changes though with [matrixed pipelinetasks](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md), maybe what you wanted to say this?

```suggestion
We will create TaskRun(s) per PipelineTask and use them as the source of truth for the status of the execution of that PipelineTask.
```

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:338]: could you please add some rationale on why these features are optional? also, can we reframe this as features that will be supported in future work instead of optional?

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:158]: sounds good - though would prefer that we make this decision (authoring time vs runtime) before marking this TEP `implementable`

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:311]: this changes though with [matrixed pipelinetasks](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md), maybe what you wanted to say this?

```suggestion
We will create TaskRun(s) per PipelineTask and use them as the source of truth for the status of the execution of that PipelineTask.
```

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [inline teps/0044-data-locality-and-pod-overhead-in-pipelines.md:337]: could you please add some rationale on why these features are optional? also, can we reframe this as features that will be supported in future work instead of optional?

**Review by jerop — COMMENTED** [review]
looks good, thanks @lbernick! 

left some comments/questions about authoring vs runtime, mapping btwn pipelinetasks and taskruns, and optional features

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:158]: definitely agree! the goal of this PR is mainly to get consensus on pipeline in a pod as a preferred solution, and I think authoring time vs runtime will be discussed in more detail when we get to the API changes.

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:339]: just FYI I did end up adding a milestone for these two features!

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (lbernick) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:341]: renamed optional -> future work! I didn't provide rationale for individual features, but I'm seeing this as a cost benefit tradeoff since this set of features isn't directly related to the data locality problem. I suspect that for matrix (for example), the benefit is high, and the implementation effort is lower, while retries will be much more challenging to implement so we may want to determine how much benefit we're providing. I didn't provide rationale for individual features as the execution mode isn't really relevant to how useful they are (just how difficult they are to implement), but I added a bit more detail to the description of the section.

**Comment by lbernick** [comment]
thanks @jerop, should be ready for another look!

**Comment by pritidesai** [comment]
API WG - @jerop to lgtm after reviewing

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:341]: thanks for updating it to future work and expounding on rationale, much clearer :)

File: teps/0044-data-locality-and-pod-overhead-in-pipelines.md
Comment (jerop) [reply teps/0044-data-locality-and-pod-overhead-in-pipelines.md:636]: yes, pipelines in pipelines is coming very soon 😁

**Review by jerop — APPROVED** [review]
looks great, thanks @lbernick 🎉

**Comment by jerop** [comment]
as agreed upon at API WG on 03/14, lgtm'ing offline after approvals from all assignees

/lgtm

**Pull request merged**
